syntax = "proto3";

package greet;

// The greeting service definition.
service Chat {
	// single call
	rpc SingleCall (ChatRequest) returns (ChatReply);

	// Server Streaming
	rpc StreamToClient (ChatRequest) returns (stream ChatReply);
	
	// Client Streaming
	rpc StreamToServer (stream ChatRequest) returns (DelayedReply);

	// Bi-directional Streaming
	rpc MutualStream (stream ChatRequest) returns (stream ChatReply);
}

// The request message containing the user's name.
message ChatRequest {
  string name = 1;
  string greeting = 2;
}

// The response message containing the greetings.
message ChatReply {
  string message = 1;
}

message DelayedReply {
	string message = 1;
	repeated ChatRequest request = 2;
}

